src/bulletproof/mod.rs

```diff
@@ -0,0 +1,848 @@
+    ///     &pc_gens,
+    ///     secret_value,
+    ///     &blinding,
+    ///     32,
+    /// ).expect("A real program could handle errors");
+    ///
+    /// assert!(
+    ///     proof
+    ///         .verify_single(&bp_gens, &pc_gens, &committed_value, 32)
+    ///         .is_ok()
+    /// );
+    /// # }
+    /// ```
+    pub fn prove_single_with_rng<T: RngCore + CryptoRng>(
+        bp_gens: &BulletproofGens,
+        pc_gens: &PedersenGens,
+        v: u64,
+        v_blinding: &Scalar,
+        n: usize,
+        rng: &mut T,
+    ) -> Result<(RangeProof, CompressedEdwardsY), ProofError> {
+        let (p, Vs) =
+            RangeProof::prove_multiple_with_rng(bp_gens, pc_gens, &[v], &[*v_blinding], n, rng)?;
+        Ok((p, Vs[0]))
+    }
+
+    /// Create a rangeproof for a given pair of value `v` and blinding
+    /// scalar `v_blinding`.
+    ///
+    /// This is a convenience wrapper around
+    /// [`RangeProof::prove_single_with_rng`], passing in a threadsafe
+    /// RNG.
+    pub fn prove_single(
+        bp_gens: &BulletproofGens,
+        pc_gens: &PedersenGens,
+        v: u64,
+        v_blinding: &Scalar,
+        n: usize,
+    ) -> Result<(RangeProof, CompressedEdwardsY), ProofError> {
+        RangeProof::prove_single_with_rng(
+            bp_gens,
+            pc_gens,
+            v,
+            v_blinding,
+            n,
+            &mut rand::thread_rng(),
+        )
+    }
+
+    /// Create a rangeproof for a set of values.
+    ///
+    /// # Example
+    /// ```
+    /// use rand::thread_rng;
+    ///
+    /// use curve25519_dalek::scalar::Scalar;
+    ///
+    /// use monero::bulletproof::{BulletproofGens, PedersenGens, RangeProof};
+    ///
+    /// # fn main() {
+    /// // Generators for Pedersen commitments.  These can be selected
+    /// // independently of the Bulletproofs generators.
+    /// let pc_gens = PedersenGens::default();
+    ///
+    /// // Generators for Bulletproofs, valid for proofs up to bitsize 64
+    /// // and aggregation size up to 16.
+    /// let bp_gens = BulletproofGens::new(64, 16);
+    ///
+    /// // Four secret values we want to prove lie in the range [0, 2^32)
+    /// let secrets = [4242344947u64, 3718732727u64, 2255562556u64, 2526146994u64];
+    ///
+    /// // The API takes blinding factors for the commitments.
+    /// let blindings: Vec<_> = (0..4).map(|_| Scalar::random(&mut thread_rng())).collect();
+    ///
+    /// // Create an aggregated 32-bit rangeproof and corresponding commitments.
+    /// let (proof, commitments) = RangeProof::prove_multiple(
+    ///     &bp_gens,
+    ///     &pc_gens,
+    ///     &secrets,
+    ///     &blindings,
+    ///     32,
+    /// ).expect("A real program could handle errors");
+    ///
+    /// assert!(
+    ///     proof
+    ///         .verify_multiple(&bp_gens, &pc_gens, &commitments, 32)
+    ///         .is_ok()
+    /// );
+    /// # }
+    /// ```
+    pub fn prove_multiple_with_rng<T: RngCore + CryptoRng>(
```

Comment:
Are all of these functions useful in the context of Monero?

Reply:
I think all of these `prove_` and `verify_` methods make sense. Perhaps the ones that don't let you control the rng can be removed or made private.

In any case, we may want to offer even more opinionated APIs for proving and verifying, and leave everything else as an implementation detail.

Reply:
> In any case, we may want to offer even more opinionated APIs for proving and verifying, and leave everything else as an implementation detail.

I think that would be good. We can always expand the API surface later. What do you think about exposing functions that directly work with the `ringct::BulletProof` type and make everything else an implementation detail? (Not on the type but simply taking it as a parameter.)